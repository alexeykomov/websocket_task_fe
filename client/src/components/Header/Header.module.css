.app-header {
  height: 48px;
  padding: 0 24px;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 20px;
  background: cornflowerblue;
  color: #fff;
}

.app-header-fixed {
  position: fixed;
  top: 0;
  width: 100%;
}

@media (max-width: 600px) {
  .app-header {
    padding: 0 12px;
  }
}

.root {
  display: flex;
  justify-content: flex-start;
  align-items: center;
  appearance: none;
  background: transparent;
  border: none;
  width: 44px;
  height: 44px;
  padding: 0;
  margin: 0;
  cursor: pointer;
  -webkit-tap-highlight-color: transparent;
}

.root:focus-visible {
  outline: none;
  box-shadow: 0 0 0 3px #fff;
}

.menu-container {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: stretch;
  width: 26px;
  height: 20px;
}

.bar {
  background: #fff;
  height: 3px;
}
