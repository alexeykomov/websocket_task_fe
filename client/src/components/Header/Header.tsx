import React from 'react';
import styles from './Header.module.css';
import clsx from 'clsx';

export interface HeaderProps {
  onMenuToggle: () => void;
  fixed: boolean;
}

export const Header = ({ onMenuToggle, fixed }: HeaderProps) => {
  return (
    <header
      className={clsx(
        styles['app-header'],
        fixed && styles['app-header-fixed']
      )}
      data-testid="app-header"
    >
      Sensors data
    </header>
  );
};
