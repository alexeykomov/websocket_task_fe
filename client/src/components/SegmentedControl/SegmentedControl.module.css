.root {
  display: flex;
  justify-content: flex-end;
  align-items: center;
  width: 100%;
}

@media (max-width: 600px) {
  .root {
    width: 100%;
  }
}

.label-cont {
  display: inline-flex;
  box-shadow: 0 0 6px rgba(0, 0, 0, 0.2);
}

@media (max-width: 600px) {
  .label-cont {
    width: 100%;
  }
}

.label {
  box-sizing: border-box;
  border: none;
  height: 48px;
  color: #323232;
  padding: 16px;
  list-style-type: none;
  text-align: center;
  border-left: #e0e0e0 1px solid;
  cursor: pointer;
  user-select: none;
  transition: background-color 0.2s ease-in-out, color 0.2s ease-in-out;
  -webkit-tap-highlight-color: transparent;
}

@media (max-width: 400px) {
  .label {
    border-left: none;
  }
}

@media (max-width: 600px) {
  .label {
    flex: 1;
  }
}

.label.last {
  border-left: none;
}

.input:checked + .label,
.input:active + .label {
  background: #16d379;
  color: #fff;
}

.input:focus-visible + .label {
  outline: none;
  box-shadow: 0 0 0 3px rgba(83, 151, 241, 0.5);
}

.input {
  position: absolute;
  width: 1px;
  height: 1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
}
