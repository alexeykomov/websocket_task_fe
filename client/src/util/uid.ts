let uid = 0;

const premappedIds = new Map();

export const getUid = () => {
  return uid++;
};

export const getUidFor = (reserved: string) => {
  if (!premappedIds.has(reserved)) {
    premappedIds.set(reserved, getUid());
  }
  return premappedIds.get(reserved);
};
