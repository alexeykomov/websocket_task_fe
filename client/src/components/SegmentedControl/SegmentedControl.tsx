import { ChangeEvent, useState } from 'react';
import { getUid } from '../../util/uid';
import styles from './SegmentedControl.module.css';
import clsx from 'clsx';
import React from 'react';
import { ConnectionFilter } from '../../features/sensor/sensorConstants';


export interface SegmentedControlProps {
  options: Array<{ label: string; value: string }>;
  onOptionSelect: (value: string) => void;
  selectedOption: string;
  testUid?: number;
}

export const SegmentedControl = ({
  options,
  onOptionSelect,
  selectedOption,
  testUid,
}: SegmentedControlProps) => {
  const [uid] = useState(Number.isInteger(testUid) ? testUid : getUid());

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    if (e.target.checked) {
      const value = e.target.value;
      onOptionSelect(value as ConnectionFilter);
    }
  };

  return (
    <div className={styles.root} data-testid={`segmented-control-${uid}`}>
      <div className={styles['label-cont']}>
        {options.map((option, index) => {
          const id = `id-${option.value}-${uid}`;
          return (
            <React.Fragment key={option.value}>
              <input
                type="radio"
                id={id}
                name={`segmented-control-${uid}`}
                value={option.value}
                className={styles.input}
                onChange={handleChange}
                checked={option.value === selectedOption}
                data-testid={`segmented-control-input-${uid}-${index}`}
              />
              <label
                htmlFor={id}
                className={clsx(
                  styles.label,
                  index === options.length - 1 && styles['label-last']
                )}
                data-testid={`segmented-control-label-${uid}-${index}`}
              >
                {option.label}
              </label>
            </React.Fragment>
          );
        })}
      </div>
    </div>
  );
};
