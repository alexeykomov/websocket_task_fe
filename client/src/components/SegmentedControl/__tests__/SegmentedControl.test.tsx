import { act, fireEvent, render } from '@testing-library/react';
import { Header } from '../../Header/Header';
import { SegmentedControl } from '../SegmentedControl';
import { getUid } from '../../../util/uid';

describe('SegmentedControl', () => {
  it('renders without crashing', () => {
    const uid = getUid();
    const { getByTestId } = render(
      <SegmentedControl
        selectedOption={'all'}
        onOptionSelect={() => {}}
        options={[]}
        testUid={uid}
      />
    );

    expect(getByTestId(`segmented-control-${uid}`)).toBeInTheDocument();
  });

  it.skip('calls onOptionSelect when an option is clicked', () => {
    const onOptionSelect = jest.fn();
    const uid = getUid();
    const { getByTestId } = render(
      <SegmentedControl
        selectedOption={'all'}
        onOptionSelect={onOptionSelect}
        options={[
          {
            label: 'All',
            value: 'all',
          },
          {
            label: 'Active',
            value: 'active',
          },
          {
            label: 'Completed',
            value: 'completed',
          },
        ]}
        testUid={uid}
      />
    );
    act(() => {
      fireEvent.change(getByTestId(`segmented-control-input-${uid}-${1}`), {
        target: {
          value: 'active',
          checked: true,
        },
      });
    });
    expect(onOptionSelect).toHaveBeenCalledWith('all');
  });

  it('matches snapshot', () => {
    const uid = 1;
    const { container } = render(
      <SegmentedControl
        selectedOption={'all'}
        onOptionSelect={() => {}}
        options={[
          {
            label: 'All',
            value: 'all',
          },
          {
            label: 'Active',
            value: 'active',
          },

          {
            label: 'Completed',
            value: 'completed',
          },
        ]}
        testUid={uid}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
