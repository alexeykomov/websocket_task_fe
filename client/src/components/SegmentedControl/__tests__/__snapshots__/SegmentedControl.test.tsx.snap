// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SegmentedControl matches snapshot 1`] = `
<div>
  <div
    class="root"
    data-testid="segmented-control-1"
  >
    <div
      class="label-cont"
    >
      <input
        checked=""
        class="input"
        data-testid="segmented-control-input-1-0"
        id="id-all-1"
        name="segmented-control-1"
        type="radio"
        value="all"
      />
      <label
        class="label"
        data-testid="segmented-control-label-1-0"
        for="id-all-1"
      >
        All
      </label>
      <input
        class="input"
        data-testid="segmented-control-input-1-1"
        id="id-active-1"
        name="segmented-control-1"
        type="radio"
        value="active"
      />
      <label
        class="label"
        data-testid="segmented-control-label-1-1"
        for="id-active-1"
      >
        Connected
      </label>
      <input
        class="input"
        data-testid="segmented-control-input-1-2"
        id="id-completed-1"
        name="segmented-control-1"
        type="radio"
        value="completed"
      />
      <label
        class="label label-last"
        data-testid="segmented-control-label-1-2"
        for="id-completed-1"
      >
        Completed
      </label>
    </div>
  </div>
</div>
`;
